### ISSUE TITLE
Configure check for JPEG extension parameters broken

### ISSUE THREAD

**Issue by felixbuenemann** [jcupitt/libvips#431]
The configure check for JPEG extension parameters, which sets `HAVE_JPEG_EXT_PARAMS` is broken, because it does not respect `JPEG_LIBS` and `JPEG_INCLUDES` set by `FIND_JPEG`:

```sh
./configure --prefix=/usr/local/Cellar/vips/DEV \
--with-jpeg-includes=/usr/local/opt/mozjpeg/include \
--with-jpeg-libraries=/usr/local/opt/mozjpeg/lib \
--enable-gtk-doc --enable-introspection
# config.log contains:
configure:26199: checking for JPEG
configure:26344: result: libraries -L/usr/local/opt/mozjpeg/lib -ljpeg, headers -I/usr/local/opt/mozjpeg/include
configure:26367: checking for JPEG extension parameters
configure:26385: gcc -c -g -O2  conftest.c >&5
conftest.c:128:5: error: use of undeclared identifier 'J_BOOLEAN_PARAM'
    J_BOOLEAN_PARAM test;
    ^
1 error generated.
```

The installed mozjpeg version is 3.1, which supports the ext params api, but configure instead uses the libjpeg from /usr/local for this test. This leads to vips being compiled with mozjpeg, but it's always using the mozjpeg defaults and none of the extended jpegsave options work.

**Comment by lovell** [comment]
Hello, I added this in https://github.com/jcupitt/libvips/commit/00e27def8ea841db6b3ec9da00bbdc976241c4ae#diff-67e997bcfdac55191033d57a16d1408a.

My best guess would be that *something* needs to be (temporarily?) modified to include the value of  `JPEG_INCLUDES` before `AC_LANG_PROGRAM` is called, but I'm not sure what that *something* is.

**Comment by felixbuenemann** [comment]
Maybe this can be merged into FIND_JPEG…

**Comment by felixbuenemann** [comment]
I mean if FIND_JPEG tries to compile with a jboolean test program first and fallback to the normal test, it could detect includes, libs and ext params.

**Comment by jcupitt** [comment]
The usual pattern is to save and restore `LIBS` or `CFLAGS` like this:

```
  save_LIBS=$LIBS
  LIBS="$LIBS $MAGICK_LIBS"
  AC_CHECK_FUNCS(ResetImageAttributeIterator,
    AC_DEFINE(HAVE_RESETIMAGEATTRIBUTEITERATOR,1,
              [define if your magick has ResetImageAttributeIterator.]))
  LIBS=$save_LIBS
```

So ugly ... but that's the way it is.

**Comment by felixbuenemann** [comment]
I tried that last evening, by setting LIBS and INCLUDES to the JPEG_* ones, but it wasn't picked up…

**Comment by jcupitt** [comment]
How about looking for `jpeg_c_bool_param_supported` instead? That's a bit of API that's only in mozjpeg, and API entry points are very simple to test for. I'll make a patch.

**Comment by felixbuenemann** [comment]
@jcupitt I'm not sure I get what you're saying. How is checking for `jpeg_c_bool_param_supported` instead of `J_BOOLEAN_PARAM` going to help here?

**Comment by felixbuenemann** [comment]
Nevermind, it works:

```
checking for JPEG... libraries -L/usr/local/opt/mozjpeg/lib -ljpeg, headers -I/usr/local/opt/mozjpeg/include
checking for jpeg_c_bool_param_supported... yes
```

**Comment by jcupitt** [comment]
It seems to work for me. If I have mozjpeg installed to `~/vips` and run vips configure like this:

```
$ CFLAGS="-g -Wall" CXXFLAGS="-g -Wall" ./configure --prefix=/home/john/vips --enable-debug
```

I see this in configure output:

```
...
checking for giflib... libraries -lgif, headers in default path
checking for PNG... yes
checking for JPEG... libraries -ljpeg, headers in default path
checking for jpeg_c_bool_param_supported... no
checking for EXIF... yes
...
```

If I run configure like this:

```
$ CFLAGS="-g -Wall" CXXFLAGS="-g -Wall" ./configure --prefix=/home/john/vips --enable-debug --with-jpeg-includes=/home/john/vips/include --with-jpeg-libraries=/home/john/vips/lib
```

I see:

```
...
checking for giflib... libraries -lgif, headers in default path
checking for PNG... yes
checking for JPEG... libraries -L/home/john/vips/lib -ljpeg, headers -I/home/john/vips/include
checking for jpeg_c_bool_param_supported... yes
checking for EXIF... yes
checking exif-data.h usability... yes
checking exif-data.h presence... yes
...
```

And I can use the new options:

```
$ vips jpegsave k2.jpg x.jpg --trellis-quant
memory: high-water mark 3.72 MB
$ 
```

**Comment by jcupitt** [comment]
Yes, ABI is very simple and safe to check for.

**Comment by felixbuenemann** [comment]
Fixed in 73080bb4a5c219154b30faa4a97ab92336064a70.

**Comment by myfreeweb** [comment]
This still does not work system-wide (without a custom prefix), e.g. even with

```
checking for JPEG... libraries -L/usr/local/lib/mozjpeg -ljpeg, headers -I/usr/local/include/mozjpeg
```

it does not define `HAVE_JPEG_EXT_PARAMS` because it uses the header from `/usr/local/include`.

(Which is libjpeg-turbo 1.5.3 here. So the comment about "JPEG extension parameters available in libjpeg-turbo >=1.5.0, mozjpeg >=3.0" seems completely wrong. I even github searched `jpeg_c_bool_param_supported` in libjpeg-turbo, there's nothing.)

This hack fixes the header:

```diff
--- libvips/foreign/Makefile.am.orig    2018-02-11 16:18:57 UTC
+++ libvips/foreign/Makefile.am
@@ -60,5 +60,5 @@ libforeign_la_SOURCES = \

 EXTRA_DIST =

-AM_CPPFLAGS = -I${top_srcdir}/libvips/include @VIPS_CFLAGS@ @VIPS_INCLUDES@
+AM_CPPFLAGS = -I${top_srcdir}/libvips/include @VIPS_INCLUDES@ @VIPS_CFLAGS@
```

```diff
--- configure.ac.orig   2018-01-03 08:25:22 UTC
+++ configure.ac
@@ -1011,12 +1011,10 @@ FIND_JPEG(

 # JPEG extension parameters available in libjpeg-turbo >=1.5.0, mozjpeg >=3.0
 if test x"$with_jpeg" != "xno"; then
-  save_LIBS="$LIBS"
-  LIBS="$LIBS $JPEG_LIBS"
+  CFLAGS="$JPEG_INCLUDES $JPEG_LIBS $CFLAGS"
   AC_CHECK_FUNCS(jpeg_c_bool_param_supported,
     AC_DEFINE(HAVE_JPEG_EXT_PARAMS,1,
              [define if your libjpeg has extension parameters.]))
-  LIBS="$save_LIBS"
 fi

 # libexif
@@ -1106,8 +1104,8 @@ do
 done | sort -ru`
 VIPS_CFLAGS=`echo $VIPS_CFLAGS`
 VIPS_CFLAGS="$VIPS_DEBUG_FLAGS $VIPS_CFLAGS"
-VIPS_INCLUDES="$ZLIB_INCLUDES $PNG_INCLUDES $TIFF_INCLUDES $JPEG_INCLUDES"
-VIPS_LIBS="$ZLIB_LIBS $MAGICK_LIBS $PNG_LIBS $TIFF_LIBS $JPEG_LIBS $GTHREAD_LIBS $REQUIRED_LIBS $EXPAT_LIBS $PANGOFT2_LIBS $GSF_LIBS $FFTW_LIBS $ORC_LIBS
 $LCMS_LIBS $GIFLIB_LIBS $RSVG_LIBS $POPPLER_LIBS $OPENEXR_LIBS $OPENSLIDE_LIBS $CFITSIO_LIBS $LIBWEBP_LIBS $LIBWEBPMUX_LIBS $MATIO_LIBS $EXIF_LIBS -lm"
+VIPS_INCLUDES="$JPEG_INCLUDES $ZLIB_INCLUDES $PNG_INCLUDES $TIFF_INCLUDES"
+VIPS_LIBS="$JPEG_LIBS $ZLIB_LIBS $MAGICK_LIBS $PNG_LIBS $TIFF_LIBS $GTHREAD_LIBS $REQUIRED_LIBS $EXPAT_LIBS $PANGOFT2_LIBS $GSF_LIBS $FFTW_LIBS $ORC_LIBS
 $LCMS_LIBS $GIFLIB_LIBS $RSVG_LIBS $POPPLER_LIBS $OPENEXR_LIBS $OPENSLIDE_LIBS $CFITSIO_LIBS $LIBWEBP_LIBS $LIBWEBPMUX_LIBS $MATIO_LIBS $EXIF_LIBS -lm"

 AC_SUBST(VIPS_LIBDIR)
```

But then there's library loading… even adding `LIBS=-Wl,-rpath,/usr/local/lib/mozjpeg` doesn't help, only `LD_LIBRARY_PATH` at runtime :(

```
$ readelf -a /usr/local/lib/libvips.so | rg RPATH
155: 0x000000000000000f (RPATH)              Library rpath: [/usr/local/lib/mozjpeg]
$ readelf -a /usr/local/bin/vips | rg RPATH 
162: 0x000000000000000f (RPATH)              Library rpath: [/usr/local/lib/mozjpeg:/usr/local/lib]
$ ldd /usr/local/lib/libvips.so | rg jpeg 
2:      libjpeg.so.8 => /usr/local/lib/libjpeg.so.8 (0x801434000)
$ vips jpegsave test.jpg test1.jpg --trellis-quant                 
/usr/local/lib/libvips.so.42: Undefined symbol "jpeg_c_int_param_supported"
$ LD_LIBRARY_PATH=/usr/local/lib/mozjpeg ldd /usr/local/lib/libvips.so | rg jpeg
2:      libjpeg.so.8 => /usr/local/lib/mozjpeg/libjpeg.so.8 (0x801434000)
$ LD_LIBRARY_PATH=/usr/local/lib/mozjpeg vips jpegsave test.jpg test1.jpg --trellis-quant                 
```

**Comment by lovell** [comment]
> So the comment about "JPEG extension parameters available in libjpeg-turbo >=1.5.0, mozjpeg >=3.0" seems completely wrong.

This comment was written prior to the release of libjpeg-turbo v1.5.0 and was based on the plan to add `jpeg_c_bool_param_supported` to it.

https://github.com/mozilla/mozjpeg/pull/165#issuecomment-87595748 provides the context.

Since then the mozjpeg maintainers have kindly added my suggestion of `#define JPEG_C_PARAM_SUPPORTED` via commit https://github.com/mozilla/mozjpeg/commit/1374cdfabeee18a83d5d7ff0358015c4bd1dec72 which would probably make a better compile-time detection method.

**Comment by jcupitt** [comment]
I tested and it seems to work for me. `jpeg_c_bool_param_supported` is defined here, it's the first thing in the extensions:

https://github.com/mozilla/mozjpeg/blob/master/jcext.c#L19

If I configure libvips 8.6 against the Ubuntu system libs (I'm just using libjpeg-turbo) I see:

```
checking for JPEG... libraries -ljpeg, headers in default path
checking for jpeg_c_bool_param_supported... no
```

And the extensions are disabled. 

If I install mozjpeg 3.2 (te current release) to ~/vips (the dir I normally install vips to) and configure libvips again, I see:

```
$ ./configure --prefix=/home/john/vips --with-jpeg-includes=/home/john/vips/include --with-jpeg-libraries=/home/john/vips/lib
... snip
checking for JPEG... libraries -L/home/john/vips/lib -ljpeg, headers -I/home/john/vips/include
checking for jpeg_c_bool_param_supported... yes
... snip
```

And I can use the extensions:

```
$ ldd $(which vips) | grep jpeg
	libjpeg.so.62 => /home/john/vips/lib/libjpeg.so.62 (0x00007fc91bd31000)
	libjpeg.so.8 => /usr/lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007fc9166d1000)
$ vips jpegsave k2.jpg x.jpg --trellis-quant
$
```

I notice at least mozjpeg now supports pkg-config (AT LAST), so we should probably use that in preference to the frankly awful `FIND_JPEG`.

**Comment by jcupitt** [comment]
Ah, thanks Lovell, I hadn't seen that issue. I shall read!

**Comment by myfreeweb** [comment]
> `--prefix=/home/john/vips`

I want to use mozjpeg in a *system wide* installation (`/usr/local` prefix). I have turbo in `/usr/local/lib` and mozjpeg in `/usr/local/lib/mozjpeg` (that's how FreeBSD packages install them).

**Comment by jcupitt** [comment]
The prefix doesn't matter, the same procedure will work anywhere. Just substitute `/usr/local` for `/home/john/vips`.

It sounds like you might have to set `LD_LIBRARY_PATH` or start experimenting with `rpath` to resolve your linking issue.

**Comment by jcupitt** [comment]
Sorry @myfreeweb, I think I may have read your original post too quickly. Let me think about this.

**Comment by lovell** [comment]
@myfreeweb My best guess would be that one of `$ZLIB_INCLUDES $PNG_INCLUDES $TIFF_INCLUDES` on your machine happens to also include the path for the libjpeg-turbo headers, which would explain why moving `$JPEG_INCLUDES` to the start of `VIPS_INCLUDES=` and therefore have a higher precedence appears to "fix" it.

**Comment by myfreeweb** [comment]
Yes, of course, all of them include `-L/usr/local/lib`.

**Comment by jcupitt** [comment]
I pushed a change to 8.6 HEAD to use pkg-config to find libjpeg, which should help a bit. I updated that comment as well.

I'm not sure what to do about the order of includes. I agree your proposed change will fix libjpeg, but won't the new order equally (potentially) break other import libraries? Perhaps the difference is that there are multiple versions of libjpeg, but (I'm praying nightly) only one libpng (for example). 

Should dependencies with more than one solution always come first in the include and library search path?
